request/request.go

```diff
@@ -128,3 +128,13 @@ func GetExtraResources(req *v1beta1.RunFunctionRequest) (map[string][]resource.E
 	}
 	return out, nil
 }
+
+func GetCredentials(req *v1beta1.RunFunctionRequest) (map[string][]map[string][]byte, error) {
+	out := make(map[string][]map[string][]byte, len(req.GetCredentials()))
+	for name, cred := range req.GetCredentials() {
+		out[name] = []map[string][]byte{}
+		data := cred.GetCredentialData().GetData()
```

Comment:
Will this function be future-proof if we add another option to the `source` oneof in the `Credentials` message? It's possible we might want to support something other than `map<string, bytes> data`.

Perhaps `GetCredentials` should return a map of string to credential struct, so we can update the struct to implement other credential types?

Reply:
Indeed, that's what I had in mind; the problem is that I am unsure how to verify the source because `GetSource` is returning `isCredentials_Source`, a private interface. If I have a condition against a string, it gives me a type error, so it seems to import [those](https://github.com/negz/crossplane/blob/8e6087c13e9998139325bc8bf5ffacbad3f73762/apis/apiextensions/v1beta1/zz_generated.composition_common.go#L334-L346) would not help.

I have updated it to return a map of credentials by converting the secret data to a map of interfaces

Reply:
It seems we're supposed to use a type switch - see https://protobuf.dev/reference/go/go-generated/#oneof

So I think something like this would work:

```go
	for name, cred := range req.GetCredentials() {
		switch cred.GetSource().(type) {
		case *v1beta1.Credentials_CredentialData:
			out[name] = Credentials{Data: cred.GetCredentialData().GetData()}
		default:
			// Unsupported. Maybe return an error?
		}
	}
```

Reply:
Thanks for the help @negz I'll have a deeper look on these docs